add open device registration test with cpp extensions

bdhirsh opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #80477
* #80418
* #80416

Adding a test for open device registration using cpp extensions.

bdhirsh committed:
add open device registration test with cpp extensions

[ghstack-poisoned]

albanD commented on test/test_cpp_extensions_open_device_registration.py:
```diff
@@ -0,0 +1,90 @@
+
+def remove_build_path():
+    if sys.platform == "win32":
+        print("Not wiping extensions build folder because Windows")
```
print?

bdhirsh replied:
I copied this from an existing cpp extension test we have, but agreed - will clean up

albanD commented on test/test_cpp_extensions_open_device_registration.py:
```diff
@@ -0,0 +1,90 @@
+
+        self.assertFalse(module.custom_add_called())
+
+        # calls out custom add kernel, registered to the dispatcher
```
nit: our

albanD commented on test/cpp_extensions/open_registration_extension.cpp:
```diff
@@ -0,0 +1,96 @@
+}
+
+
+TORCH_LIBRARY_IMPL(aten, PrivateUse1, m) {
```
nit: add a comment that explains what this is doing?

bdhirsh replied:
agreed

albanD commented on test/cpp_extensions/open_registration_extension.cpp:
```diff
@@ -0,0 +1,96 @@
+  m.impl("_copy_from", &custom__copy_from);
+}
+
+c10::Device get_custom_device() {
```
nit why is this needed? Can it be done directly in python?

albanD commented on test/cpp_extensions/open_registration_extension.cpp:
```diff
@@ -0,0 +1,96 @@
+  return called;
+}
+
+PYBIND11_MODULE(TORCH_EXTENSION_NAME, m) {
```
nit: comment explaining why we need this.

albanD commented on test/test_cpp_extensions_open_device_registration.py:
```diff
@@ -0,0 +1,90 @@
+        # check that it was called
+        self.assertTrue(module.custom_add_called())
+
+        z_cpu = z.to(device='cpu')
```
nit: test that the "cross device" copy did copy the values properly?

albanD reviewed (commented):
Looks great!
I think the second part of the c++ file could use similar comments to the first part of it. Added comments inline to that end.

bdhirsh committed:
Update on "add open device registration test with cpp extensions"


Adding a test for open device registration using cpp extensions.

[ghstack-poisoned]

bdhirsh committed:
Update on "add open device registration test with cpp extensions"


Adding a test for open device registration using cpp extensions.

[ghstack-poisoned]

bdhirsh committed:
Update on "add open device registration test with cpp extensions"


Adding a test for open device registration using cpp extensions.

[ghstack-poisoned]

bdhirsh committed:
Update on "add open device registration test with cpp extensions"


Adding a test for open device registration using cpp extensions.

[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge

bdhirsh committed:
Update on "add open device registration test with cpp extensions"


Adding a test for open device registration using cpp extensions.

[ghstack-poisoned]

bdhirsh committed:
Update on "add open device registration test with cpp extensions"


Adding a test for open device registration using cpp extensions.

[ghstack-poisoned]

albanD commented on test/cpp_extensions/open_registration_extension.cpp:
```diff
@@ -0,0 +1,114 @@
+// More details on the dispatcher can be found at http://blog.ezyang.com/2020/09/lets-talk-about-the-pytorch-dispatcher/.
+TORCH_LIBRARY_IMPL(aten, PrivateUse1, m) {
+  m.impl("add.Tensor", &custom_add_Tensor);
+  m.impl("empty.memory_format", &custom_empty_memory_format);
```
This needs to be empty.SymInt now

albanD reviewed (approved):
SGTM thanks

bdhirsh committed:
Update on "add open device registration test with cpp extensions"


Adding a test for open device registration using cpp extensions.

[ghstack-poisoned]

bdhirsh commented:
@pytorchbot help

bdhirsh commented:
@pytorchbot --help

bdhirsh commented:
@pytorchbot rebase

bdhirsh committed:
Update on "add open device registration test with cpp extensions"


Adding a test for open device registration using cpp extensions.

[ghstack-poisoned]

bdhirsh committed:
Update on "add open device registration test with cpp extensions"


Adding a test for open device registration using cpp extensions.

[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge
